### PR TITLE
PullRequest for Issue31: added support for AMDSScaler

### PR DESCRIPTION
**Pull request by XuetaoLiu-S**
Add supports for scalers to retrieve the scaler data

### CODE REVIEW

File: source/DataElement/AMDSFlatArray.cpp [hunk 1]
```diff
@@ -200,6 +200,106 @@ AMDSFlatArray AMDSFlatArray::operator /(const quint32 divisor)
 	return newDataFlatArray;
 }
 
+bool AMDSFlatArray::setValues(QVector<quint8> values)
+{
+	if (dataType_ != AMDSDataTypeDefinitions::Unsigned8 || size() > values.size()) {
+		return false;
+	}
+
+	vectorQuint8_ = QVector<quint8>(values);
```
Comment (dretrex) [inline source/DataElement/AMDSFlatArray.cpp]: Why is this not just `vectorQuint8_ = values;`?

File: source/DataElement/AMDSFlatArray.h [hunk 2]
```diff
@@ -17,6 +18,18 @@ class AMDSFlatArray
 	/// define the Division operation of AMDSFlatArray: the value of the given handler will be divided by the given divisior
 	virtual AMDSFlatArray operator /(quint32 divisor);
 
+	/// set the values of the data with the given vector
+	bool setValues(QVector<quint8> values) ;
```
Comment (dretrex) [inline source/DataElement/AMDSFlatArray.h]: These should be `const QVector<> &`

File: source/DataHolder/AMDSGenericFlatArrayDataHolder.cpp [hunk 3]
```diff
@@ -3,6 +3,7 @@
 #include "DataHolder/AMDSDataHolderSupport.h"
 #include "util/AMErrorMonitor.h"
 
+#include <QDebug>
```
Comment (dretrex) [inline source/DataHolder/AMDSGenericFlatArrayDataHolder.cpp]: Leftover include.

File: source/DataHolder/AMDSScalarDataHolder.cpp [hunk 4]
```diff
@@ -1,6 +1,7 @@
 #include "AMDSScalarDataHolder.h"
 
 
+#include <QDebug>
```
Comment (dretrex) [inline source/DataHolder/AMDSScalarDataHolder.cpp]: Here too.

File: source/Detector/Scaler/AMDSScalerConfigurationMap.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+#include "DataElement/AMDSDataTypeDefinitions.h"
+
+class AMDSScalerConfigurationMap : public QObject
+{
+    Q_OBJECT
+public:
+	explicit AMDSScalerConfigurationMap(const QString &scalerName, const QString &basePVName, AMDSDataTypeDefinitions::DataType dataType, const QList<quint8> &enabledChannelId, QObject *parent = 0);
+	~AMDSScalerConfigurationMap();
+
+	inline const QString scalerName() { return scalerName_; }
```
Comment (dretrex) [inline source/Detector/Scaler/AMDSScalerConfigurationMap.h]: `const` functions.

File: source/Detector/Scaler/AMDSScalerDetector.cpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	connected_ = false;
+	AMErrorMon::alert(this, 0, "PV control is NOT connected.");
+}
+
+void AMDSScalerDetector::onDwellTimeControlValueChanged(double newValue)
+{
+	if (dwellTime_ != newValue) {
+		AMErrorMon::alert(this, 0, QString("Scaler %1 dwell time changed from %2ms to %3ms.").arg(scalerName()).arg(dwellTime_).arg(newValue));
+
+		dwellTime_ = (int)newValue;
```
Comment (dretrex) [inline source/Detector/Scaler/AMDSScalerDetector.cpp]: This should be casted before the comparison.  That will ensure better non-floating point equivalent comparisons.

File: source/Detector/Scaler/AMDSScalerDetectorManager.cpp [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+	connect(detectorManagerThread_, SIGNAL(finished()), scalerDetector_, SLOT(deleteLater()));
+
+	detectorManagerThread_->start();
+}
+
+AMDSScalerDetectorManager::~AMDSScalerDetectorManager()
+{
+	if (detectorManagerThread_->isRunning())
+		detectorManagerThread_->terminate();
```
Comment (dretrex) [inline source/Detector/Scaler/AMDSScalerDetectorManager.cpp]: This should be `quit()` shouldn't it?

File: source/Detector/Scaler/AMDSScalerDetectorManager.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+class AMDSScalerDetector;
+
+class AMDSScalerDetectorManager : public QObject
+{
+    Q_OBJECT
+public:
+	explicit AMDSScalerDetectorManager(AMDSScalerConfigurationMap *scalerConfiguration, QObject *parent = 0);
+	~AMDSScalerDetectorManager();
+
+	inline AMDSScalerDetector *scalerDetector() { return scalerDetector_; }
```
Comment (dretrex) [inline source/Detector/Scaler/AMDSScalerDetectorManager.h]: `const` function.

**Comment by dretrex** [comment]
Only the couple comments.
Time Estimate: 10 minutes

File: source/DataElement/AMDSFlatArray.cpp (hunk 1 above)
Comment (XuetaoLiu-S) [inline source/DataElement/AMDSFlatArray.cpp]: nice caught :+1: . I made a cast when I was testing.

**Pull request merged**